cpp/src/reductions/reductions.cu

```diff
@@ -38,231 +35,177 @@ void gpu_reduction_op(const T *data, const gdf_valid_type *mask,
+    template <typename T, typename U>
+    static constexpr bool is_convertable()
+    {
+        return std::is_same<T, U>::value ||
+            ( std::is_arithmetic<T>::value && std::is_arithmetic<U>::value );
     }
 
-    template<typename T>
-    static constexpr T identity() { return T{1}; }
-};
-
-struct DeviceSumOfSquares {
-    struct Loader {
-        template<typename T>
-        __device__
-        T operator() (const T* ptr, int pos) const {
-            T val = ptr[pos];   // load
-            return val * val;   // squared
-        }
-    };
-    // round 2 just uses the basic sum reduction
-    typedef DeviceSum second;
-
-    template<typename T>
-    __device__
-    T operator() (const T &lhs, const T &rhs) const {
-        return lhs + rhs;
+public:
+    template <typename T_out, typename std::enable_if<
+                is_convertable<T_in, T_out>() >::type* = nullptr >
+    gdf_error operator()(const gdf_column *col, void *dev_result)
+    {
+        return ReduceOp<T_in, T_out, Op>
+            (col, static_cast<T_out*>(dev_result));
     }
 
-    template<typename T>
-    static constexpr T identity() { return T{0}; }
-};
-
-struct DeviceMin {
-    typedef IdentityLoader Loader;
-    typedef DeviceMin second;
-
-    template<typename T>
-    __device__
-    T operator() (const T &lhs, const T &rhs) {
-        return lhs <= rhs? lhs: rhs;
-    }
-
-    template<typename T>
-    static constexpr T identity() { return std::numeric_limits<T>::max(); }
-};
-
-struct DeviceMax {
-    typedef IdentityLoader Loader;
-    typedef DeviceMax second;
-
-    template<typename T>
-    __device__
-    T operator() (const T &lhs, const T &rhs) {
-        return lhs >= rhs? lhs: rhs;
+    template <typename T_out, typename std::enable_if<
+                !is_convertable<T_in, T_out>() >::type* = nullptr >
+    gdf_error operator()(const gdf_column *col, void *dev_result)
+    {
+        return GDF_UNSUPPORTED_DTYPE;
     }
-
-    template<typename T>
-    static constexpr T identity() { return std::numeric_limits<T>::lowest(); }
 };
 
 template <typename Op>
 struct ReduceDispatcher {
-    template <typename T,
-              typename std::enable_if_t<std::is_arithmetic<T>::value>* = nullptr>
-    gdf_error operator()(gdf_column *col, 
-                         void *dev_result, 
-                         gdf_size_type dev_result_size) {
+private:
+    // return true if T is arithmetic type or
+    // Op is DeviceMin or DeviceMax for wrapper (non-arithmetic) types
+    template <typename T>
+    static constexpr bool is_supported()
+    {
+        return std::is_arithmetic<T>::value ||
+               std::is_same<Op, DeviceMin>::value ||
+               std::is_same<Op, DeviceMax>::value ;
+    }
+
+public:
+    template <typename T, typename std::enable_if<
+        is_supported<T>()>::type* = nullptr>
+    gdf_error operator()(const gdf_column *col,
+                         void *dev_result, gdf_dtype output_dtype)
```

Comment:
Add a `stream` parameter defaulted to `0` to every API below `gdf_reduction`. This will make it easier in the future when expose streams in the external API.